### ISSUE TITLE
NotSerializableException for Metadata

### ISSUE THREAD

**Issue by jakekdodd** [DigitalPebble/storm-crawler#87]
Getting `NotSerializableException: com.digitalpebble.storm.crawler.Metadata` when running CrawlTopology on the master branch. For reference, I'm running the topology in local mode using maven exec.

I'm 99.99% sure that Metadata just needs to implement Serializable (this solves the problem for me). Just wanted to check if CrawlTopology was successfully executing for anyone else after the most recent commits.

Stack trace:

```
15352 [Thread-17-disruptor-executor[6 6]-send-queue] ERROR backtype.storm.util - Async loop died!
java.lang.RuntimeException: java.lang.RuntimeException: java.io.NotSerializableException: com.digitalpebble.storm.crawler.Metadata
	at backtype.storm.utils.DisruptorQueue.consumeBatchToCursor(DisruptorQueue.java:128) ~[storm-core-0.9.3.jar:0.9.3]
	at backtype.storm.utils.DisruptorQueue.consumeBatchWhenAvailable(DisruptorQueue.java:99) ~[storm-core-0.9.3.jar:0.9.3]
	at backtype.storm.disruptor$consume_batch_when_available.invoke(disruptor.clj:80) ~[storm-core-0.9.3.jar:0.9.3]
	at backtype.storm.disruptor$consume_loop_STAR_$fn__1460.invoke(disruptor.clj:94) ~[storm-core-0.9.3.jar:0.9.3]
	at backtype.storm.util$async_loop$fn__464.invoke(util.clj:463) ~[storm-core-0.9.3.jar:0.9.3]
	at clojure.lang.AFn.run(AFn.java:24) [clojure-1.5.1.jar:na]
	at java.lang.Thread.run(Thread.java:745) [na:1.7.0_55]
Caused by: java.lang.RuntimeException: java.io.NotSerializableException: com.digitalpebble.storm.crawler.Metadata
	at backtype.storm.serialization.SerializableSerializer.write(SerializableSerializer.java:41) ~[storm-core-0.9.3.jar:0.9.3]
	at com.esotericsoftware.kryo.Kryo.writeClassAndObject(Kryo.java:568) ~[kryo-2.21.jar:na]
	at com.esotericsoftware.kryo.serializers.CollectionSerializer.write(CollectionSerializer.java:75) ~[kryo-2.21.jar:na]
	at com.esotericsoftware.kryo.serializers.CollectionSerializer.write(CollectionSerializer.java:18) ~[kryo-2.21.jar:na]
	at com.esotericsoftware.kryo.Kryo.writeObject(Kryo.java:486) ~[kryo-2.21.jar:na]
	at backtype.storm.serialization.KryoValuesSerializer.serializeInto(KryoValuesSerializer.java:44) ~[storm-core-0.9.3.jar:0.9.3]
	at backtype.storm.serialization.KryoTupleSerializer.serialize(KryoTupleSerializer.java:44) ~[storm-core-0.9.3.jar:0.9.3]
	at backtype.storm.daemon.worker$mk_transfer_fn$fn__3549.invoke(worker.clj:129) ~[storm-core-0.9.3.jar:0.9.3]
	at backtype.storm.daemon.executor$start_batch_transfer__GT_worker_handler_BANG_$fn__3283.invoke(executor.clj:258) ~[storm-core-0.9.3.jar:0.9.3]
	at backtype.storm.disruptor$clojure_handler$reify__1447.onEvent(disruptor.clj:58) ~[storm-core-0.9.3.jar:0.9.3]
	at backtype.storm.utils.DisruptorQueue.consumeBatchToCursor(DisruptorQueue.java:125) ~[storm-core-0.9.3.jar:0.9.3]
	... 6 common frames omitted
Caused by: java.io.NotSerializableException: com.digitalpebble.storm.crawler.Metadata
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1183) ~[na:1.7.0_55]
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:347) ~[na:1.7.0_55]
	at backtype.storm.serialization.SerializableSerializer.write(SerializableSerializer.java:38) ~[storm-core-0.9.3.jar:0.9.3]
	... 16 common frames omitted
15353 [Thread-17-disruptor-executor[6 6]-send-queue] ERROR backtype.storm.daemon.executor - 
java.lang.RuntimeException: java.lang.RuntimeException: java.io.NotSerializableException: com.digitalpebble.storm.crawler.Metadata
	at backtype.storm.utils.DisruptorQueue.consumeBatchToCursor(DisruptorQueue.java:128) ~[storm-core-0.9.3.jar:0.9.3]
	at backtype.storm.utils.DisruptorQueue.consumeBatchWhenAvailable(DisruptorQueue.java:99) ~[storm-core-0.9.3.jar:0.9.3]
	at backtype.storm.disruptor$consume_batch_when_available.invoke(disruptor.clj:80) ~[storm-core-0.9.3.jar:0.9.3]
	at backtype.storm.disruptor$consume_loop_STAR_$fn__1460.invoke(disruptor.clj:94) ~[storm-core-0.9.3.jar:0.9.3]
	at backtype.storm.util$async_loop$fn__464.invoke(util.clj:463) ~[storm-core-0.9.3.jar:0.9.3]
	at clojure.lang.AFn.run(AFn.java:24) [clojure-1.5.1.jar:na]
	at java.lang.Thread.run(Thread.java:745) [na:1.7.0_55]
Caused by: java.lang.RuntimeException: java.io.NotSerializableException: com.digitalpebble.storm.crawler.Metadata
	at backtype.storm.serialization.SerializableSerializer.write(SerializableSerializer.java:41) ~[storm-core-0.9.3.jar:0.9.3]
	at com.esotericsoftware.kryo.Kryo.writeClassAndObject(Kryo.java:568) ~[kryo-2.21.jar:na]
	at com.esotericsoftware.kryo.serializers.CollectionSerializer.write(CollectionSerializer.java:75) ~[kryo-2.21.jar:na]
	at com.esotericsoftware.kryo.serializers.CollectionSerializer.write(CollectionSerializer.java:18) ~[kryo-2.21.jar:na]
	at com.esotericsoftware.kryo.Kryo.writeObject(Kryo.java:486) ~[kryo-2.21.jar:na]
	at backtype.storm.serialization.KryoValuesSerializer.serializeInto(KryoValuesSerializer.java:44) ~[storm-core-0.9.3.jar:0.9.3]
	at backtype.storm.serialization.KryoTupleSerializer.serialize(KryoTupleSerializer.java:44) ~[storm-core-0.9.3.jar:0.9.3]
	at backtype.storm.daemon.worker$mk_transfer_fn$fn__3549.invoke(worker.clj:129) ~[storm-core-0.9.3.jar:0.9.3]
	at backtype.storm.daemon.executor$start_batch_transfer__GT_worker_handler_BANG_$fn__3283.invoke(executor.clj:258) ~[storm-core-0.9.3.jar:0.9.3]
	at backtype.storm.disruptor$clojure_handler$reify__1447.onEvent(disruptor.clj:58) ~[storm-core-0.9.3.jar:0.9.3]
	at backtype.storm.utils.DisruptorQueue.consumeBatchToCursor(DisruptorQueue.java:125) ~[storm-core-0.9.3.jar:0.9.3]
	... 6 common frames omitted
Caused by: java.io.NotSerializableException: com.digitalpebble.storm.crawler.Metadata
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1183) ~[na:1.7.0_55]
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:347) ~[na:1.7.0_55]
	at backtype.storm.serialization.SerializableSerializer.write(SerializableSerializer.java:38) ~[storm-core-0.9.3.jar:0.9.3]
	... 16 common frames omitted
Disconnected from the target VM, address: '127.0.0.1:49418', transport: 'socket'
15370 [Thread-17-disruptor-executor[6 6]-send-queue] ERROR backtype.storm.util - Halting process: ("Worker died")
java.lang.RuntimeException: ("Worker died")
	at backtype.storm.util$exit_process_BANG_.doInvoke(util.clj:325) [storm-core-0.9.3.jar:0.9.3]
	at clojure.lang.RestFn.invoke(RestFn.java:423) [clojure-1.5.1.jar:na]
	at backtype.storm.daemon.worker$fn__3808$fn__3809.invoke(worker.clj:452) [storm-core-0.9.3.jar:0.9.3]
	at backtype.storm.daemon.executor$mk_executor_data$fn__3274$fn__3275.invoke(executor.clj:240) [storm-core-0.9.3.jar:0.9.3]
	at backtype.storm.util$async_loop$fn__464.invoke(util.clj:473) [storm-core-0.9.3.jar:0.9.3]
	at clojure.lang.AFn.run(AFn.java:24) [clojure-1.5.1.jar:na]
	at java.lang.Thread.run(Thread.java:745) [na:1.7.0_55]

Process finished with exit code 1
```

**Comment by jnioche** [comment]
You are not using the [ConfigurableTopology](https://github.com/DigitalPebble/storm-crawler/blob/master/core/src/main/java/com/digitalpebble/storm/crawler/ConfigurableTopology.java#L58)?
This adds `Config.registerSerialization(conf, Metadata.class);` to the config, which is needed for the Metadata to work.

**Comment by jakekdodd** [comment]
CrawlTopology extends ConfigurableTopology, so at first glance the serialization registration should be taken care of. I'll take a closer look.

**Comment by jakekdodd** [comment]
There it is...the registration only happens when not running in local mode.
```java
 if (isLocal) {
            LocalCluster cluster = new LocalCluster();
            cluster.submitTopology(name, conf, builder.createTopology());
        }

        else {
            // register Metadata for serialization with FieldsSerializer
            Config.registerSerialization(conf, Metadata.class);
            try {
                StormSubmitter.submitTopology(name, conf,
                        builder.createTopology());
            } catch (Exception e) {
                e.printStackTrace();
                return -1;
            }
        }
```

**Comment by jnioche** [comment]
Sorry read your email too quickly, should have seen that you were using
CrawlTopology. Weird, I thought the serialization happened only in
distributed mode. Good catch! do you want to patch it or shall I?
Is there another place we should put this in, in case people don't use
ConfigurableTopology?


On 5 February 2015 at 14:43, Jake Dodd <notifications@github.com> wrote:

> There it is...the registration only happens when not running in local mode.
>
>  if (isLocal) {
>             LocalCluster cluster = new LocalCluster();
>             cluster.submitTopology(name, conf, builder.createTopology());
>         }
>
>         else {
>             // register Metadata for serialization with FieldsSerializer
>             Config.registerSerialization(conf, Metadata.class);
>             try {
>                 StormSubmitter.submitTopology(name, conf,
>                         builder.createTopology());
>             } catch (Exception e) {
>                 e.printStackTrace();
>                 return -1;
>             }
>         }
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/DigitalPebble/storm-crawler/issues/87#issuecomment-73146132>
> .
>



-- 



* Open Source Solutions for Text Engineering   http://www.digitalpebble.com
<http://www.digitalpebble.com>*
*http://digitalpebble.blogspot.com <http://digitalpebble.blogspot.com/>*
https://twitter.com/digitalpebble

**Comment by jakekdodd** [comment]
Yeah I'll look into moving the registration somewhere else and then patch it...at the very least, we can document the fact that a user must manually register Metadata for serialization if they choose to not extend ConfigurableTopology

**Comment by jakekdodd** [comment]
Patched in 2cb21606d5fae54670922d9897d1938f4388803a.

Will make a new issue for documenting the Metadata registration when not using ConfigurableTopology
